### PR TITLE
[controls] fix memory leak in Style and AttachedCollection

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/Vroomer/MAUI-master-detail-memory-leak
Context: https://github.com/symbiogenesis/Maui.DataGrid/tree/memory-leak
Context: https://github.com/dotnet/maui/issues/12039

Using the Visual Studio's `.NET Object Allocation Tracking` profiler
on the sample above I noticed after using the app and taking snapshots:

Snapshot 2:

    WeakReference<Microsoft.Maui.Controls.BindableObject>
    Count: 1,545
    Bytes: 37,080

Snapshot 10:

    WeakReference<Microsoft.Maui.Controls.BindableObject>
    Count: 1,657
    Bytes: 39,768

It appears memory usage is slowly growing by about ~2.45KB each time I
navigate forward a page then return.

f8171b4 introduced a `CleanupTrigger`, but the code has a slight issue:

    void CleanUpWeakReferences()
    {
        if (_targets.Count < _cleanupThreshold)
        {
            return;
        }

        _targets.RemoveAll(t => !t.TryGetTarget(out _));
        _cleanupThreshold = _targets.Count + CleanupTrigger;
    }

Because `_cleanupThreshold` is modified at the bottom of this method,
the `_targets` collection increases by 128 in size over time. The
result being lots of tiny `WeakReference<T>` objects in memory, but
not the actual underlying `T` target.

I did a little research to try to find a better collection than
`List<WeakReference<T>>`:

* https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2
* https://source.dot.net/#Microsoft.Extensions.Caching.Memory/MemoryCache.cs
* https://source.dot.net/#PresentationFramework/System/Windows/Input/KeyboardNavigation.cs,08271c29723aa10a

I did not find one that exactly met our needs, so I wrote a very
simple `WeakList<T>`. It cleans up `WeakReference<T>` that are no
longer alive after N operations. I set this threshold to 32 to start,
we can modify this later if needed. This makes the above samples call
it a couple times for large `CollectionView`'s, which seems reasonable.

After these changes, I'm able to navigate in the app over and over
with stable memory usage:

| ID | Objects | Heap Size     |
| -- | --:     | --:           |
| 1  | 107,885 | 407,645.87 KB |
| 2  |   2,267 | 401,126.22 KB |

This was after navigating through the app *many* times, and it
actually went *down*.

Then I found `Microsoft.Maui.Internals.WeakList<T>`, which I did not
like quite as much as my implementation. I deleted/replaced its usage
adding new unit tests for `WeakList<T>`.

Note that I believe I found one other leak, so this doesn't fully
solve #12039 yet.

### CODE REVIEW

**Comment by jonathanpeppers** [comment]
Placed you'd hit this in apps:

* You have a `<DataTemplate/>` using a `Style` -- the default template has many built-in styles
* You scroll the `CollectionView`/`ListView`
* You navigate forward & back to pages with these

**Comment by hartez** [comment]
The idea was that we'd increase the threshold if there were a lot of legitimate items still in memory, so that we weren't scanning that list of references all the time without anything to clean up. 

Obviously since we never shrink the threshold again, we get stuck with a bigger and bigger number of references sitting around. So either the threshold needs to be occasionally reset, or just thrown out entirely.

**Comment by jonathanpeppers** [comment]
I think a second issue was the collection would only "evict" things on `Style.Apply()`.

It seems good to move the logic to a new class, write some unit tests for it, and test a couple more user samples to see if it is fixed.

My changes here would also evict `WeakReference<T>` objects during `GetEnumerator()` and `Remove()` calls.

File: src/Core/tests/UnitTests/WeakListTests.cs [hunk 1]
```diff
@@ -18,7 +18,8 @@ public async Task ObjectsAreEvicted_GetEnumerator()
 			GC.Collect();
 			GC.WaitForPendingFinalizers();
 
-			foreach (var _ in list) ;
+			foreach (var _ in list)
+				;
```
Comment (rmarinho) [inline src/Core/tests/UnitTests/WeakListTests.cs:22]: hmm can this be express on editorconfig ?

Comment (jonathanpeppers) [reply src/Core/tests/UnitTests/WeakListTests.cs:22]: If someone else wants to fix that, I'll probably just put a `{ }` block with a comment inside.

**Comment by jonathanpeppers** [comment]
So, I don't think this fully solves https://github.com/dotnet/maui/issues/12039.

I still see these objects hanging around in some cases (some apps):

https://github.com/dotnet/maui/blob/255c435d2d4a1fcb19d241e650f7688a788f915f/src/Controls/src/Core/BindingExpression.cs#L609

Let me solve that one in a separate PR, I believe I see an issue to solve.

File: src/Core/src/HotReload/HotReloadHelper.cs [hunk 2]
```diff
@@ -155,7 +154,7 @@ static void RegisterHandler(KeyValuePair<Type, Type> pair, Type newHandler)
 
 		public static void TriggerReload()
 		{
-			List<IHotReloadableView?>? roots = null;
+			List<IHotReloadableView>? roots = null;
```
Comment (jonathanpeppers) [inline src/Core/src/HotReload/HotReloadHelper.cs:157]: The old `WeakList` was a `IList<T?>`, I removed nullability.

**Review by hartez — APPROVED** [review]
This looks great, and potentially cleaning up on GetEnumerator() seems like a good idea.

**Review by mattleibow — APPROVED** [review]

**Review by rmarinho — APPROVED** [review]

**Comment by kojini** [comment]
I don't see whether this fix will be backported to .NET 7.  
Will it be backported to .NET 7? Do you have ETA on it?
(FYI, I'll be posting the same questions regarding some of memory leak issues/PRs as I'm tracking down when the bug fixes might be released)

**Pull request merged**